### PR TITLE
More portable C flags

### LINKED ISSUE #11426: [configure] float precision flags are C compiler-dependent.
As pointed out in the discussion at https://github.com/coq/coq/pull/9867#issuecomment-575791940 , different C compilers do need different flags regarding floating point support.

Configure should thus detect the C compiler OCaml is using `config.c_compiler` [note that this doesn't have to be a single command]

There is also the question if another flag could be used, for example `-std=c99`.

### PR DESCRIPTION
**Pull request by proux01**
Use -std=c99 instead of the GCC argument -fexcess-precision=standard
This requires the -fasm as the VM is using the asm GNU extension
(also implemented by other compilers).

These flags should be more portable accross C compilers.

Fixes #11426

### CODE REVIEW

**Review by ejgallego — APPROVED** [review]
Good for me, it needs review from the VM / kernel maintainers.

**Comment by ppedrot** [comment]
@silene ping

**Comment by ejgallego** [comment]
I confirm the build with these flags is warning-free with `gcc` 9.2.1 , I dunno about `clang`

**Comment by ejgallego** [comment]
@maximedenes you were going to comment here, but you couldn't.

IMHO as the old configure maintainer and with my C hat on this patch is OK. We discussed this in the call and the discussion was a bit long; but my impression is that the main worry what about the effects this could have on the VM codebase.

I think there is not a lot to worry, C99 is widely accepted and used, and declaring our code compliant is a good step.

Thus I propose to merge in a few days.

**Comment by ppedrot** [comment]
Somebody should probably send an email to @silene though, just in case.

**Comment by ejgallego** [comment]
Not sure having to send individual emails for nit changes such as this is going to scale, also I don't like that part of the discussion data would not be stored where it belongs. [Also I just saw @silene online on GH]

This is kinda trivial, proposed by the maintainer of the feature, correct, fixes a bug, and open for more than one month so IMHO people had plenty of time to discuss; would we have any problem people can report and request a revert.

**Comment by maximedenes** [comment]
Is the OCaml VM compiled with the same flags?

**Comment by maximedenes** [comment]
Maybe @bgregoir can have a quick look and tell us what he thinks.

**Comment by ppedrot** [comment]
@ejgallego I am convinced that messing with the C flags is far from trivial. As discussed during the call, this could even interfere with the C-implemented OCaml runtime.

**Comment by ejgallego** [comment]
Sorry folks I'm unable to understand the concerns here; I'll let somebody else handle this.

**Comment by proux01** [comment]
> Is the OCaml VM compiled with the same flags?

Yes (and I really don't see how we could avoid it).

**Comment by proux01** [comment]
> @ejgallego I am convinced that messing with the C flags is far from trivial. As discussed during the call, this could even interfere with the C-implemented OCaml runtime.

Sure but in this case, the added flags mostly make the compiler accept less programs. It's not impossible that something would get compiled differently but we would be terribly unlucky.

If some worry remain, maybe could we ask OCaml folks their opinion about compiling the OCaml VM with those flags?

**Comment by maximedenes** [comment]
> Yes (and I really don't see how we could avoid it).

Do you have a pointer?

I see https://github.com/ocaml/ocaml/blob/bb15443a237d69716d97e6e7f96f4377dde83a74/configure.ac#L545 which seems to say they use `-std=gnu99` for GCC instead. Also, https://github.com/ocaml/ocaml/blob/bb15443a237d69716d97e6e7f96f4377dde83a74/configure.ac#L565 seems to suggest they don't pass `-std=c99` to clang (they write "Clang defaults to gnu99 or gnu11, which is fine").

So I guess the naive question is why we don't take the same `-std=gnu99` flag as OCaml.

**Comment by proux01** [comment]
`gnu99` means `c99` + GNU extensions. Since `asm` is the only GNU extension used, it should then be equivalent to `-std=c99 -fasm`

> So I guess the naive question is why we don't take the same -std=gnu99 flag as OCaml.

We could. However, if we want to exactly copy OCaml and pass it only to GCC and not to Clang, this would require aditional code in the configure to determine the compiler.

Should I do it? (this would add a few extra lines in configure.ml)

**Comment by ejgallego** [comment]
> Should I do it?

Please don't introduce compiler-dependent configuration options, specially when the weaker flags do work.

**Comment by maximedenes** [comment]
> Since `asm` is the only GNU extension used, it should then be equivalent to `-std=c99 -fasm`

I see.

> Should I do it? (this would add a few extra lines in configure.ml)

I don't think that's necessary.

**Review by silene — APPROVED** [review]
The semantics differences between `c99` and `gnu99` are irrelevant (on the C99 subset). So, if the VM compiles with `-std=c99`, all the better.

**Comment by maximedenes** [comment]
@silene Thanks for the feedback! The failures seem unrelated, so merging.

@proux01 Thanks for your patience!

**Comment by proux01** [comment]
You're welcome.

Anyway, the discussion wasn't useless, it raises an interesting point : according to the comment you pointed out (https://github.com/ocaml/ocaml/blob/bb15443a237d69716d97e6e7f96f4377dde83a74/configure.ac#L540 and the following code), OCaml is compiled with `O2` without `strict-aliasing` and with `wrapv` whereas Coq VM is just compiled with `O2` (since this commit https://github.com/coq/coq/commit/2887393cf817d0509caf7a2bb8f7850e2bc2d123). Shouldn't we imitate OCaml? I'm opening a separate PR.

BTW @ppedrot , I'm more worried by those last two flags than this `c99` flag.

**Comment by silene** [comment]
Good point. That said, I am not too worried about `-fno-strict-aliasing` and  `-fwrapv`. As far as I know, there is no instance of type punning in Coq's VM, and all the integers are unsigned, contrarily to OCaml.

**Comment by ejgallego** [comment]
It seems that OCaml master has recently become incompatible with `std=c99` due to its use of `sigset_t` which is not part of the standard; I dunno what should we do here.

It seems the right thing to do is to use `-D_XOPEN_SOURCE=600` as to increase the POSIX level support, but I guess that should be done in OCaml proper, cc: @xavierleroy (not sure who the right person to ping is, sorry if you are not)

**Comment by proux01** [comment]
> It seems the right thing to do is to use -D_XOPEN_SOURCE=600 as to increase the POSIX level support, but I guess that should be done in OCaml proper

Isn't OCaml compiled with `-std=gnu99` ? doesn't this subsume this `_XOPEN_SOURCE=600` ?

**Comment by ejgallego** [comment]
> Isn't OCaml compiled with `-std=gnu99` ?

It seems it is not.

> doesn't this subsume this `_XOPEN_SOURCE=600` ?

It seems std99 doesn't imply POSIX2004 indeed, current OCaml master is using something more lax and thus Coq seems to break unless we up our POSIX level. We should hear from OCaml devs tho, maybe we can open an issue?

**Comment by proux01** [comment]
> > Isn't OCaml compiled with `-std=gnu99` ?
> 
> It seems it is not.

Well, from https://github.com/ocaml/ocaml/blob/bb15443a237d69716d97e6e7f96f4377dde83a74/configure.ac#L545 (comment and code below), I thought they were using either `gnu99` or `gnu11`

> > doesn't this subsume this `_XOPEN_SOURCE=600` ?
> 
> It seems std99 doesn't imply POSIX2004 indeed, current OCaml master is using something more lax and thus Coq seems to break unless we up our POSIX level. We should hear from OCaml devs tho, maybe we can open an issue?

What happens if we just add this `-D_XOPEN_SOURCE=600` to Coq cflags?

**Comment by ejgallego** [comment]
Oh sorry, I misread gnu99 for std99.

Indeed it is compiled with `gnu` not with `std`, which does indeed work in this case.

I'm not sure if we should add that flag, or just follow what upstream does flag-wise. The problem here comes from OCaml's own `.h` files not working in our setting.

**Comment by proux01** [comment]
Indeed, switching to `gnu99`, mimicking OCaml, might avoid such issues in the future.

**Comment by xavierleroy** [comment]
I didn't know Coq was playing with the `-std=xxx` C compiler options.  

What you're seeing is a collision between two facts:
* With (gcc `\/` clang) `/\` glibc, at least, `-std=c99` means "strict C99 compatibility", i.e. "the standard headers declare exactly what's specified in ISO C99 and nothing more".  It does not  just mean "please let me use C99 features that are not in C90". 
* OCaml's support for pre-POSIX, pure-ISO-C signal handling was a bunch of `#ifdef` that were never exercised (except Windows, all our platforms are POSIX, and Windows has special signal handling code anyway), so it made sense to remove it and demand that the C compiler and standard libraries implement POSIX signals.

As a consequence, OCaml can no longer be compiled with `-std=c99`, just like most Unix programs cannot be compiled with `-std=c99`.  

Now my question: why does Coq play with the `-std=` options of the C compiler rather than using the defaults?  That could be GCC's defaults or OCaml's defaults (if you compile .c files using ocamlc or ocamlopt), and that would work better than guessing your own options.

**Comment by ejgallego** [comment]
Hi @xavierleroy , thanks for the comments

> Now my question: why does Coq play with the `-std=` options of the C compiler rather than using the defaults?

We enabled `--std=c99` due to the floating point support in Coq; the discussion above points to the original bug, but the short story is that Coq needs `-fexcess-precision=standard` which seems not supported by `clang`, so we opted for `--std=c99` which implies it.

> That could be GCC's defaults or OCaml's defaults (if you compile .c files using ocamlc or ocamlopt), and that would work better than guessing your own options.

Unfortunately in this case the defaults don't suffice, as `--std=gnu99` doesn't seem to be added; what would you recommend?

**Comment by proux01** [comment]
> That could be GCC's defaults or OCaml's defaults (if you compile .c files using ocamlc or ocamlopt), and that would work better than guessing your own options.

Indeed, we should do that, thanks for the advice!

> Unfortunately in this case the defaults don't suffice, as --std=gnu99 doesn't seem to be added?

Probably because it's already the default (or gnu11) for your compiler, it would be added for gcc <= 4 (c.f. https://github.com/ocaml/ocaml/blob/bb15443a237d69716d97e6e7f96f4377dde83a74/configure.ac#L545).

**Comment by ejgallego** [comment]
I see, so indeed we should just remove both flags as already OCaml guarantees >= `gnu99`.

I got indeed a bit confused [the perks of abandoning C some years ago] w.r.t. gcc defaults, indeed modern gcc sets `--std=gnu11`, and `--std=gnu99` after gcc 5 I think. `clang` seems to use `gnu17` by default so that should be good I assume?

**Comment by proux01** [comment]
That's also my understanding.

**Comment by ejgallego** [comment]
PR submitted.

**Comment by xavierleroy** [comment]
FYI: based on my experiments in the context of https://github.com/ocaml/ocaml/issues/7917 , it looks like `gcc -std=gnu99` does NOT imply `-fexcess-precision=standard`.  

Don't freak out: I'm looking for ways to add `-fexcess-precision=standard` to the options that OCaml passes to GCC, and will report here when it's working.

**Comment by proux01** [comment]
> FYI: based on my experiments in the context of [ocaml/ocaml#7917](https://github.com/ocaml/ocaml/issues/7917) , it looks like `gcc -std=gnu99` does NOT imply `-fexcess-precision=standard`.

So `gnu99` would not mean "`c99` plus a bunch of extensions" but rather "`c99` plus a bunch of extensions minus a few things"?

> Don't freak out: I'm looking for ways to add `-fexcess-precision=standard` to the options that OCaml passes to GCC, and will report here when it's working.

@xavierleroy This may be a stupid question but how does this compare to sticking to `c99` and selectively adding a few extensions (`-fasm`, `-D_XOPEN_SOURCE=600` and maybe a few others)?

**Pull request merged**
